### PR TITLE
Use glogger as remote logging infra

### PR DESCRIPTION
**Pull request by Jongy**
## Description

Closes: https://github.com/Granulate/gprofiler/issues/525

I didn't support this feature in A/B mode as suggested in the ticket, it's much changes than I expected and it'll be hard to maintain both infras in the code.

Some changes I did as part of it:
* state `run_id` and `cycle_id` are no longer added in logging level, then omitted in local logger - I add them in glogger level.

TODOs:
* [x] Call the handler.close() method of glogger - is it needed to ensure all logs are sent back? Or does glogger have some kind of finalization?

## How Has This Been Tested?

* [x] `NO_SERVER_LOG_KEY` doesn't send the message to the server.
* [x] `NO_SERVER_EXTRA_KEY` doesn't send `extra`s to the server.
* [x] `NO_SERVER_LOG_KEY` and `NO_SERVER_EXTRA_KEY` are not sent to server / printed locally.
* [x] plain messages are logged locally
* [x] plain messages marked with `NO_SERVER_LOG_KEY` are logged locally
* [x] messages with extras are logged locally with the extras.
* [x] messages with extras marked with `NO_SERVER_EXTRA_KEY` are logged locally with the extras.
* [x] messages with traceback logged locally with traceback
* [x] messages with traceback logged remotely with traceback
* [x] `cycle_id` passes to the backend properly
* [x] changing `--glogger-server` works

### CODE REVIEW

File: gprofiler/log.py [hunk 1]
```diff
@@ -89,101 +41,49 @@ class RemoteLogsHandler(logging.Handler):
...
+
+        return super().emit(record)
+
+    def get_metadata(self) -> Dict[str, str]:
+        from gprofiler.metadata.system_metadata import get_hostname_or_none
+
+        state = get_state()
+        hostname = get_hostname_or_none()
+
+        metadata = {
```
Comment (Jongy) [inline gprofiler/log.py:66]: I kept the same set of metadatas as we had in the previous version. Will add more on demand.

File: gprofiler/log.py [hunk 2]
```diff
@@ -89,101 +41,49 @@ class RemoteLogsHandler(logging.Handler):
...
+            "run_id": state.run_id,
+            "cycle_id": state.cycle_id,
+            "gprofiler_version": __version__,
+            "service_name": self._service_name,
         }
```
Comment (d3dave) [inline gprofiler/log.py:71]: are you sure this is supposed to be part of metadata and not extra? the metadata is shared across all sent logs at time of sending, so its possible for a batch to contain logs from separate runs/cycles.

Comment (Jongy) [reply gprofiler/log.py:71]: You're right. Good catch

Comment (Jongy) [reply gprofiler/log.py:71]: Using `get_extra`.

File: gprofiler/log.py [hunk 3]
```diff
@@ -89,101 +50,50 @@ class RemoteLogsHandler(logging.Handler):
...
-        return {
-            "message": record.message,
-            "level": record.levelname,
-            "timestamp": formatted_timestamp,
-            "extra": extra,
-            "logger_name": record.name,
-            "exception": exception_traceback,
-            RUN_ID_KEY: run_id,
-            CYCLE_ID_KEY: cycle_id,
+        if self.get_extra_fields(record).pop(NO_SERVER_EXTRA_KEY, False):
```
Comment (d3dave) [inline gprofiler/log.py:64]: calling `get_extra_fields` twice

**Comment by d3dave** [comment]
reviewed all. just that one comment.

**Review by d3dave — DISMISSED** [review]

**Comment by Jongy** [comment]
Waiting on https://github.com/Granulate/gprofiler/pull/621 so CI is fixed.

**Comment by Jongy** [comment]
I need to fix the tests (some tests referenced the `extra`s via the old way)

**Review by d3dave — APPROVED** [review]

**Comment by Jongy** [comment]
@d3dave - about this note:

>  Call the handler.close() method of glogger - is it needed to ensure all logs are sent back? Or does glogger have some kind of finalization?

How does glogger handle it? It seems that messages are sent by some mechanism, before the process goes down.

**Comment by d3dave** [comment]
> How does glogger handle it? It seems that messages are sent by some mechanism, before the process goes down.

`logging` registers an `atexit` handler that closes all stray handlers. You only need to do it if you need to flush before process exit.

**Comment by Jongy** [comment]
Made one final change - ready to merge now

**Review by d3dave — APPROVED** [review]

**Pull request merged**